lib/styleToFilters.js

```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
           layers[layerName].minzoom = style.layers[i].minzoom || 0;
           layers[layerName].maxzoom = style.layers[i].maxzoom || 22;
         }
+
+        // Collect the used property
+        layers[layerName].properties = layers[layerName].properties || [];
+        ['paint', 'layout'].forEach(item => {
+          let itemObject = style.layers[i][item];
+          itemObject && getPropertyLoop(itemObject, layers[layerName].properties);
+        });
       }
     }
   }
 
+  // remove duplicate propertys and fix choose all the propertys in layers[i].properties
+  Object.keys(layers).forEach(layerId => {
+    let properties = layers[layerId].properties;
+    if (properties.indexOf(true) !== -1) {
+      layers[layerId].properties = true;
+    } else {
+      let unique = {};
+      properties.forEach(function(i) {
+        if (!unique[i]) {
+          unique[i] = true;
+        }
+      });
+      layers[layerId].properties = Object.keys(unique);
+    }
+  })
+
   return layers;
 }
 
+
+function getPropertyLoop(propertyObj, properties) {
+  Object.keys(propertyObj).forEach(key => {
+    let value = propertyObj[key];
+    // TODO we still have outher situation:
+    // - special properties: `mapbox_clip_start`, `mapbox_clip_end`
+    if (typeof value === 'string') {
+      // if the value is string try to get property name from `xx{PropertyName}xx` like.
+      // the /{[^}]+}/ig return all the value like {xxx}
+      // eg 'a{hello}badfa' => ['{hello}']
+      // eg 'a{hello}ba{world}dfa' => ['{hello}','{world}']
+      let preProperties = value.match(/{[^}]+}/ig);
+      preProperties && preProperties.forEach(item => {
```

Comment:
Let's is this style, which I think is more readable:

```
if (preProperties) {
    preProperties.forEach ....
}
```

Reply:
emmm, I am thinking for every possible types variable to go through `a && ...` and `if(a){ //...}`, but I don't find any difference between them in my mind.
But for writing "semantic" code, I'll change to `if`.

Reply:
Yes, just `semantic`. Both ways are 💯 valid, I'm just stating my preference for `if (a) { //... }`